### PR TITLE
Special status message

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Percent encode the status message. Addresses https://github.com/grpc/grpc-dotnet/issues/161.

I have tested locally with a custom client but I'll see if I can run the interop tests as well.

### CODE REVIEW

File: test/FunctionalTests/Infrastructure/GrpcTestFixture.cs [hunk 1]
```diff
@@ -74,5 +77,17 @@ public void Dispose()
...
             _server.Dispose();
         }
+
+        public void AssertTrailerStatus(StatusCode statusCode, string details)
+        {
+            Assert.AreEqual(statusCode.ToTrailerString(), TrailersContainer.Trailers[GrpcProtocolConstants.StatusTrailer].ToString());
+            Assert.AreEqual(Uri.EscapeDataString(details), TrailersContainer.Trailers[GrpcProtocolConstants.MessageTrailer].ToString());
+        }
+
+        public void AssertSuccessTrailerStatus()
```
Comment (JamesNK) [inline test/FunctionalTests/Infrastructure/GrpcTestFixture.cs]: Change AssertSuccessTrailerStatus to call AssertTrailerStatus

File: test/FunctionalTests/Infrastructure/GrpcTestFixture.cs [hunk 2]
```diff
@@ -74,5 +77,17 @@ public void Dispose()
             Client.Dispose();
             _server.Dispose();
         }
+
+        public void AssertTrailerStatus(StatusCode statusCode, string details)
+        {
+            Assert.AreEqual(statusCode.ToTrailerString(), TrailersContainer.Trailers[GrpcProtocolConstants.StatusTrailer].ToString());
```
Comment (JamesNK) [inline test/FunctionalTests/Infrastructure/GrpcTestFixture.cs]: Nice. Now there is a central method for comparing the status code you can fix something that has been annoying me.

Update this AreEqual to include a message for when they aren't equal. Right now you get a message like "1 does not equal 13". Convert the trailer value back to the enum - `(StatusCode)Convert.ToInt32(trailerValue)` - and have a message like "Expected grpc-status Ok but got InternalError".

File: test/FunctionalTests/CompressionTests.cs [hunk 3]
```diff
@@ -240,8 +234,8 @@ public async Task SendCompressedMessageAndReturnResultWithNoCompressFlag_Respons
 
             var responseMessage = MessageHelpers.AssertReadMessage<HelloReply>(await response.Content.ReadAsByteArrayAsync().DefaultTimeout());
             Assert.AreEqual("Hello World", responseMessage.Message);
-
-            Assert.AreEqual(StatusCode.OK.ToTrailerString(), Fixture.TrailersContainer.Trailers[GrpcProtocolConstants.StatusTrailer].ToString());
+            Fixture.AssertSuccessTrailerStatus();
+            Fixture.AssertSuccessTrailerStatus();
```
Comment (JamesNK) [inline test/FunctionalTests/CompressionTests.cs]: Double success?!!?!?!!!!

**Comment by JamesNK** [comment]
@JunTaoLuo Push your changes

File: src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs [hunk 4]
```diff
@@ -205,7 +204,17 @@ public static void SetStatusTrailers(HttpResponse response, Status status)
...
             SetTrailer(response, GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
-            SetTrailer(response, GrpcProtocolConstants.MessageTrailer, status.Detail);
+
+            var escapedDetail = status.Detail;
+            if (!string.IsNullOrEmpty(status.Detail))
+            {
+                // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+                // The value portion of Status-Message is conceptually a Unicode string description of the error,
+                // physically encoded as UTF-8 followed by percent-encoding.
+                escapedDetail = Uri.EscapeDataString(status.Detail);
```
Comment (jtattermusch) [inline src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs]: qq about performance: if original the string doesn't contain any characters that would need to be percent-encoded,  is the same original string object returned?

Comment (JunTaoLuo) [reply src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs]: Yup: https://github.com/dotnet/corefx/blob/a6f76f4f620cbe74821c6445af3f13e048361658/src/System.Private.Uri/src/System/UriExt.cs#L586-L587. The EscapeString method returns null if nothing is escaped.

**Comment by jtattermusch** [comment]
Despite my initial checks, it seems that this PR has broken the  "status_code_and_message" dart interop test: https://github.com/grpc/grpc/issues/18843

https://source.cloud.google.com/results/invocations/9e8262dd-80e9-46e3-a655-f1ec5f437da6/targets/github%2Fgrpc%2Finterop_test/tests

I'm going to revert his PR for now to unbreak the tests and we can merge again after fixing the problem (it seems that we are currently performing percent-encoding overly aggressively, compared to other languages).
```
Received incorrect status: gRPC Error (2, test%20status%20message).
#0      Tester.statusCodeAndMessage (package:interop/src/client.dart:1016:9)
```

**Pull request merged**